modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/pagemem/SpeedBasedThrottleBreakdownTest.java

```diff
@@ -0,0 +1,109 @@
+import org.junit.Test;
+
+import static org.apache.ignite.cluster.ClusterState.ACTIVE;
+
+/**
+ * Tests for the cases when high load might break down speed-based throttling protection.
+ *
+ * @see PagesWriteSpeedBasedThrottle
+ */
+public class SpeedBasedThrottleBreakdownTest extends GridCommonAbstractTest {
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String gridName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(gridName);
+
+        DataStorageConfiguration dbCfg = new DataStorageConfiguration()
+                .setDefaultDataRegionConfiguration(new DataRegionConfiguration()
+                        // tiny CP buffer is required to reproduce this problem easily
+                        .setCheckpointPageBufferSize(3_000_000)
+                        .setPersistenceEnabled(true))
+                .setCheckpointFrequency(200)
+                .setWriteThrottlingEnabled(true);
+
+        cfg.setDataStorageConfiguration(dbCfg);
+
+        cfg.setConsistentId(gridName);
+
+        return cfg;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        stopAllGrids();
+
+        super.beforeTest();
+
+        deleteWorkFiles();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        stopAllGrids();
+
+        super.afterTest();
+
+        deleteWorkFiles();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected long getTestTimeout() {
+        return 3L * 60 * 1000;
+    }
+
+    /**
+     * There was the following bug: at the very start of a checkpoint, the checkpoint progress information was not yet
+     * available to the speed-based throttler, in which case it did not apply throttling. If during that short period
+     * there was high pressure on the Checkpoint Buffer, the buffer would be exhausted causing node failure.
+     *
+     * It was easy to reproduce with a small Checkpoint Buffer. This test does exactly this: with a small
+     * Checkpoint Buffer, it creates a spike of load on it. The test is successful if the CP Buffer protection
+     * does not break down and no exception gets thrown as a result.
+     *
+     * @throws Exception if something goes wrong.
+     */
+    @Test
+    public void speedBasedThrottleShouldNotAllowCPBufferBreakdownWhenCPBufferIsSmall() throws Exception {
+        Ignite ignite = startGrids(1);
+
+        ignite.cluster().state(ACTIVE);
+        IgniteCache<Object, Object> cache = ignite.createCache(DEFAULT_CACHE_NAME);
+
+        for (int i = 0; i < 100_000; i++) {
+            cache.put("key" + i, ThreadLocalRandom.current().nextDouble());
+        }
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    private void deleteWorkFiles() throws Exception {
+        cleanPersistenceDir();
+        U.delete(U.resolveWorkDirectory(U.defaultWorkDirectory(), "snapshot", false));
```

Comment:
Why is it necessary to delete the snapshot directory? I don't see them being created.